Add theming documentation. (#13)
EvoEsports/EvoSC-sharp-documentation: docs/development/modules/themes/creating-themes.md

# Creating Themes
Every theme is defined by a class which then provides options and component replacements. Here we will look into how one can create a new Theme within a module.

## How Themes Work

> When a theme class is defined, EvoSC# will automatically add this theme to the theme manager. In this process, if the theme is unique to the other themes, it is also activated. It is possible to have multiple themes active at the same time, with each, implementing their own set of options and component replacements.

Reviewer: ```suggestion
When a theme class is defined, EvoSC# will automatically add this theme to the theme manager. In this process, if the theme is unique to the other themes, it is also activated. It is possible to have multiple themes active at the same time, with each implementing their own set of options and component replacements.
```

> *This avoid conflicts within the same module, and also helps understanding what the option is used for and where it belongs to. Note that in certain cases, such as having general purpose options for a module, one can omit this.*

- Don't postfix options with redundant names such as "Color".
> *Because these options are almost always used for colors, we can avoid verbose names such as "TextColor" or "BgColor", and simply call them "Text" and "Bg".*

- Option names should be as simple and descriptive as possible, typically one word of it's primary usage.
> *For example: "Text", "Bg" for BackGround, "Border", "Divider" etc.*

- Postfix option names with their purpose if they have secondary usages, so that the name is PrimaryusageSecondaryusage.
> *For example, let's say you have a secondary text color, you can call it TextSecondary. Other examples are BgHeaderGrad1, BgHeaderGrad2, BgContent, BgRow etc.*

### Using Base Theme Options
The constructor of a theme class allows Dependency Injection. This means that we can inject the Current Theme Object, use existing options for our theme. It allows us to use the base theme to style our Manialink, in order to keep a consistent style.

> You may be wondering, can't we just use the base theme options directly instead of creating a theme with new options that uses them? This is where the philosophy of the controller comes in. Because we wan't provide people with a highly customizable and versatile server controller so that they can make it exactly how they want. So everything within a Manialink has it's own option, which can be overriden in the config or by other Modules to modify the look of it.

Reviewer: ```suggestion
You may be wondering, can't we just use the base theme options directly instead of creating a theme with new options that uses them? This is where the philosophy of the controller comes in. We want to provide people with a highly customizable and versatile server controller so that they can make it exactly how they want. Everything within a Manialink has it's own options, which can be overridden in the config or by other modules to modify the look of it.
```

```cs
[Theme(Name = "MyTheme", Description = "This is my custom theme.")]
public class MyTheme : Theme<MyTheme>
{
    public override Task ConfigureAsync()
    {
        Replace("UI.Button").With("MyModule.MyButton");

        return Task.CompletedTask;
    }
}
```

The way this work now, is that whenever `UI.Button` is imported or referenced in some way, it will actually use `MyModule.MyButton` instead.

### Overriding Existing Themes
A theme can also override existing themes using the `Override` option in the `[Theme]` attribute.

So let's say we want to override the theme for the Button component. We can then do the following:

```cs
[Theme(Name = "NewButtonTheme", Description = "New theme for the button.", Override = typeof(DefaultButtonTheme))]
public class MyTheme : DefaultButtonTheme
{
    public override async Task ConfigureAsync()
    {
        await base.ConfigureAsync();

        Set("UI.Button.Default.Bg").To("0000FF"); // set button background to blue
    }
}
```

> This shows another feature of the Theme Framework. We can inherit other themes. This is very useful if you just want to replace certain options, and not having to re-define every single option the original theme provides.

Reviewer: ```suggestion
This shows another feature of the Theme framework. We can inherit other themes. This is very useful if you just want to replace certain options without having to re-define every single option the original theme provides.
```

Author: "Theme Framework" is inteded, thats the name of it.
